Site Editor: add preview options component

vindl opened the pull request:
## Description

Integrates the PreviewOptions component from the post editor in site editor context.

I extracted the `UseResizeCanvas` and `PreviewOptions` component to `block-editor` package and it now only contains the devices menu, while the `Preview in new tab` can be optionally added in contexts where it makes sense, like the post editor. 

Part of https://github.com/WordPress/gutenberg/issues/19253.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested with Twenty Twenty theme.

- Verify that this is not causing regressions in the existing post editor functionality.
- Test the preview options in site editor.

## Screenshots <!-- if applicable -->

<img width="1680" alt="Screenshot 2020-04-16 at 17 20 23" src="https://user-images.githubusercontent.com/1182160/79479144-9ec69a00-800c-11ea-882c-008b3c1e7a7c.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo reviewed (commented):
(edit: whooops!  I missed the `in progress` tag and just starting going with the review request 😆 )

There are 2 things that catch my eye testing:

1.  These changes seem to make the 'Preview in a new tab' link disabled in places they used to be usable.  This example is taken from the standard editor for one of my pages:
![Screen Shot 2020-04-01 at 7 51 04 PM](https://user-images.githubusercontent.com/28742426/78197279-d7824300-7452-11ea-84cb-c1b395d72725.png)
Looking at the code, I'm not quite sure what would cause this.

2. I'm not sure if we are missing some necessary styles in the site editor to highlight the previews, but in the standard editor we see the gray background:
<img src="https://user-images.githubusercontent.com/28742426/78197333-06001e00-7453-11ea-9171-822b21e7be5c.png" width="200px" />

This gray background isn't apparent in the site editor, which makes it difficult to determine where the preview starts:

<img src="https://user-images.githubusercontent.com/28742426/78197423-4790c900-7453-11ea-9d11-bef60dad9e3b.png" width="200px" />

I can't speak to what packages which things belong in, but these changes seem to make sense so far.

vindl commented:
> (edit: whooops! I missed the in progress tag and just starting going with the review request 😆 )

That's fine, I'm mostly looking for reviews to see if the approach makes sense before diving into the more details.

> These changes seem to make the 'Preview in a new tab' link disabled in places they used to be usable. 

I noticed that too and at this point I'm not sure either.

> I'm not sure if we are missing some necessary styles in the site editor to highlight the previews, but in the standard editor we see the gray background:

I don't think that was the case initially in my testing but I can reproduce it now. Maybe related to some of the newer changes that I picked up with the rebase.

vindl committed:
Site Editor: add preview options component

Move PreviewOptions and useResizeCanvas to block-editor package.

vindl committed:
Refactor classnames

vindl committed:
Fix menu item padding

vindl committed:
Use gray background for site editor interface skeleton

vindl committed:
Add external preview button in edit-post context

vindl committed:
Fix bottom padding in edit-post context

vindl commented:
After chatting about this with @youknowriad I changed the approach and extracted the relevant components and moved them to `block-editor` package.

> These changes seem to make the 'Preview in a new tab' link disabled in places they used to be usable. This example is taken from the standard editor for one of my pages:

@Addison-Stavlo this should be resolved with latest changes.

> I'm not sure if we are missing some necessary styles in the site editor to highlight the previews, but in the standard editor we see the gray background:

The gray background has also been added now.

vindl commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -74,14 +76,25 @@ function Header() {
-					forceIsAutosaveable={ hasActiveMetaboxes }
-					forcePreviewLink={ isSaving ? null : undefined }
-				/>
-				<PostPreviewButton
```
This is the old preview button and I couldn't see the use for it here since it seems to be hidden. It would be nice if someone with more experience with this code could check it out and make sure that this removal won't break something.

vindl replied:
Thanks @tellthemachines, I'll look into bringing it back.

vindl replied:
@tellthemachines I brought the old preview button back along with the required styles in dcbaf85d0da3b2d90cfce7110cd19e8c925fad3b.

tellthemachines commented on packages/block-editor/src/style.scss:
```diff
@@ -36,6 +36,7 @@
 @import "./components/media-placeholder/style.scss";
 @import "./components/multi-selection-inspector/style.scss";
 @import "./components/plain-text/style.scss";
+@import "./components/preview-options/style.scss";
```
This should be moved to beneath line 55, under the marker for resizable styles, so that its contents aren't impacted by the device preview style manipulation.

vindl replied:
Updated in dcbaf85d0da3b2d90cfce7110cd19e8c925fad3b.

tellthemachines reviewed (changes requested):
Thanks for working on this! I left a couple of comments below. Also, you'll need to pop into the e2e tests and change the classnames used for targeting this component to match your changes 🙂

simison commented on packages/block-editor/src/components/preview-options/index.js:
```diff
@@ -0,0 +1,96 @@
+import { __ } from '@wordpress/i18n';
+import { check } from '@wordpress/icons';
+
+const downArrow = (
```
Is this same as https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/arrow-down.js ?

vindl replied:
No, that one looks different, you can check it out [here](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library) (also the `path` value was different in that code compared to what we have here).

youknowriad replied:
So this  is  `chevronDown`  right?

vindl replied:
> So this is chevronDown right?

Oh right, missed that one. I could pass it to icon prop of the button component, but that would place it on the left side of the text, instead of right where it's currently place. What do you think? Btw. for some reason when I replace the existing `downArrow` with `chvronDown` here nothing gets rendered. 🤔

vindl committed:
Fix e2e tests

vindl committed:
Bring back old preview button for mobile screen sizes

vindl committed:
Fix e2e tests

vindl commented:
>  Also, you'll need to pop into the e2e tests and change the classnames used for targeting this component to match your changes 🙂

@tellthemachines the e2e tests are fixed now. This should be ready for another pass. :)

Addison-Stavlo commented:
Trying this now, im having the issue that the entire background for the site editor has turned gray.

![Screen Shot 2020-04-16 at 9 15 43 AM](https://user-images.githubusercontent.com/28742426/79460589-0b1cab80-7fc3-11ea-9a70-34a44d758043.png)

![Screen Shot 2020-04-16 at 9 16 00 AM](https://user-images.githubusercontent.com/28742426/79460598-0d7f0580-7fc3-11ea-948a-4307544ddb18.png)

vindl commented:
@Addison-Stavlo I can't reproduce this, the background is the same for me as indicated in the screenshot in the summary.

youknowriad commented on packages/block-editor/src/components/index.js:
```diff
@@ -60,6 +60,8 @@ export { default as withColorContext } from './color-palette/with-color-context'
 export { default as __experimentalBlockSettingsMenuFirstItem } from './block-settings-menu/block-settings-menu-first-item';
 export { default as __experimentalInserterMenuExtension } from './inserter-menu-extension';
+export { default as __experimentalPreviewOptions } from './preview-options';
+export { default as __experimentalUseResizeCanvas } from './resize-canvas';
```
What's the reasoning for including these components and their state in the block-editor package?

vindl replied:
I was just looking for a good place for them so that they could be used both in `edit-post` and `edit-site` context. I didn't want to place them in `editor` since it depends on wp and post, and this could be useful outside of that context too. Do you think they should be placed in some other package?

vindl replied:
@youknowriad based on our conversation I removed the required state from `block-editor` component and placed it in `edit-post` and `edit-site` separately in dc50eb9cb1d3f60414aab4c4aab30f3ec97dc3b4.

I considered using local state for it, but since both `PreviewOptions` and `useResizeCanvas` rely on it, and these are at different nesting levels and parts of the editor, this seemed cleaner that propagating this through props or using context.

vindl committed:
Move device type store out of block-editor package

vindl committed:
Fix e2e tests

youknowriad commented on packages/block-editor/src/components/preview-options/index.js:
```diff
@@ -0,0 +1,87 @@
+						>
+							{ __( 'Mobile' ) }
+						</MenuItem>
+					</MenuGroup>
```
I wonder this component should  just render the MenuGroup  and not the whole dropdown, I guess you did it this way because you wanted to share dropdown code but it does seem like it reduces the semantic value of the component a bit. I don't see this as a blocker or anything for now, just something to think about.

vindl replied:
>  I guess you did it this way because you wanted to share dropdown code but it does seem like it reduces the semantic value of the component a bit.

Right, I didn't want to duplicate that code, but also it's somewhat coupled with `useResizeCanvas` now in terms of what device strings it sets in the store. Until that is reworked somehow, I think it makes sense to keep it as is.

youknowriad commented on packages/block-editor/src/components/preview-options/style.scss:
```diff
@@ -40,29 +39,12 @@
+	.block-editor-post-preview__dropdown {
 		display: flex;
 	}
 }
```
This component seem  to have a lot of custom styles, It looks very similar to "More Menus" or "Block Settings Menus"  so I wonder why it needs all these styles and can't just rely on the default UI components styles. Maybe our UI components styles are not well thought in this situation.

vindl replied:
My aim here was to preserve the current way it looks in `edit-post` completely. I think we could consider removing these, but I also thing it would be better to tackle that in a separate PR.

youknowriad commented on packages/block-editor/src/components/resize-canvas/index.js:
```diff
@@ -1,20 +1,21 @@
-		return select( 'core/edit-post' ).__experimentalGetPreviewDeviceType();
-	}, [] );
-
+export default function useResizeCanvas( deviceType ) {
```
Should this be  `use-resize-canvas` in a "hooks" folder instead of "components"?

vindl replied:
I'm not sure but I can look into it more. I placed it in components since that's where it was in `edit-post`.

youknowriad replied:
yeah weird for me, we do have "hooks" folder in some packages.

vindl replied:
I tried to rework it here as suggested 9b90de0fb5bfd32908fc847493a9b83c438b0b4e, not sure if it's the right way to go about it though. :)

youknowriad commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -81,9 +89,29 @@ function Header( { onToggleInserter, isInserterOpen } ) {
+				>
+					<MenuGroup>
+						<div className="edit-post-header-preview__grouping-external">
+							<PostPreviewButton
```
Nit: Potentially, we could extract this to a small component somewhere ini edit-post to avoid having to add selectors... specific to this component here.

vindl replied:
I think this would make sense for the whole `PreviewOptions` section here too. 🤔

vindl replied:
Refactored in 2376bfd662f1f3265b8416d137d7581bf610f0a2.

youknowriad reviewed (commented):
Left a few questions but this is looking good for me.

vindl committed:
Move useResizeCanvas to hooks

youknowriad commented on packages/block-editor/src/hooks/use-resize-canvas.js:
```diff
@@ -68,3 +70,9 @@ export function useResizeCanvas() {
 	return contentInlineStyles( deviceType );
 }
+
+addFilter(
```
Is this a new filter, did we have it before?

vindl replied:
Yes, it's a new filter.

youknowriad replied:
We've been avoiding adding new filters in Gutenberg as we figured that filters and hooks are not the best API for extensibility in JS. 

Could you clarify why this filter is needed and do you think we should have it added in its own PR instead? We could discuss alternatives... for your use cases.

vindl replied:
I don't think it's needed, it's just that I noticed that that's how hooks are usually organized/used (haven't worked with them so far in Gutenberg), so I figured that a request to move this to hooks folder would imply a filter.

youknowriad replied:
Ohhh, I guess we have a naming conflicts (react hooks vs WP hooks) 🤔

vindl replied:
But now I realize that my assumption has been wrong 😅. I pushed a revert for that in 09cfe0e673c9362d78ce409cb5f186a7abb8d1ef.

vindl committed:
Isolate edit-post device preview chunk as a separate component

vindl committed:
Replace downArrow with chevronDown icon

vindl committed:
Revert "Move useResizeCanvas to hooks"

This reverts commit 9b90de0fb5bfd32908fc847493a9b83c438b0b4e.

tellthemachines reviewed (commented):
This is looking good!

For some weird reason, the popover on the site editor is now opening towards the right, whereas on the post editor it still opens to the left. Might be worth setting an explicit `position` value in the Dropdown component so it is consistent in both places. (How it looks on post editor seems less ambiguous as to what its parent is.)
<img width="365" alt="Screen Shot 2020-04-24 at 1 20 38 pm" src="https://user-images.githubusercontent.com/8096000/80173559-946f5600-8633-11ea-9617-0f8ee733e685.png">

youknowriad reviewed (approved):
Nice

vindl committed:
Set dropdown position to bottom left

vindl commented:
> For some weird reason, the popover on the site editor is now opening towards the right, whereas on the post editor it still opens to the left. Might be worth setting an explicit position value in the Dropdown component so it is consistent in both places. (How it looks on post editor seems less ambiguous as to what its parent is.)

Good catch @tellthemachines, I updated this in 35b9bdfd4cf89f501e990e4dfcbb1af9a9478b69.
